### PR TITLE
openssl: support OpenSSL 3.0 and above

### LINKED ISSUE #1160: add OpenSSL 3.0+ support
Now that OpenSSL 1.1.1 is scarce in the wild, we should start updating the wrappers and stdlib to OpenSSL 3.0 at minimum to make sure tooling using `std/net` runs without additional packages on most *nix.

It should be safe to drop OpenSSL <1.1.1 support, as those are no longer widely used and does not support the latest TLSv1.3 standard.

We should drop LibreSSL support for now as well, as we don't have the resource to test against that library, and platforms that use LibreSSL also supports OpenSSL.

This is more or less a short term goal to get existing tools running well.

## Useful References

- [PEP 644](https://peps.python.org/pep-0644/): Python decision to only support 1.1.1 and above.

### PR DESCRIPTION
**Pull request by alaviss**
## Summary
The only ABI change between version 1.1.1 and 3.0 for our usage is 
`SSL_get_peer_certificate`  being splitted into 
`SSL_get1_peer_certificate`  (which is compatible with the prior symbol)
and  `SSL_get0_peer_certificate` .

This PR modifies  `SSL_get_peer_certificate`  in the wrapper to use the
new symbol if available. No changes in other programs are required.

## Details
*  `SSL_get_peer_certificate`  will now select either 
`SSL_get1_peer_certificate`  or  `SSL_get_peer_certificate`  depending
on which symbols are available.
* DLL names for OpenSSL 3.x has been added for macOS, Windows and
Linux.
* The symbols used for certificate verification are no longer
unconditionally hidden on Windows. They were hidden previously as Nim
ships old OpenSSL 1.0 which did not have these symbols.

Fixes https://github.com/nim-works/nimskull/issues/1160

### CODE REVIEW

File: lib/wrappers/openssl.nim [hunk 1]
```diff
@@ -796,10 +796,14 @@ when defined(nimHasStyleChecks):
 
 # Certificate validation
 # On old openSSL version some of these symbols are not available
-when not defined(nimDisableCertificateValidation) and not defined(windows):
-
-  proc SSL_get_peer_certificate*(ssl: SslCtx): PX509{.cdecl, dynlib: DLLSSLName,
-      importc.}
+when not defined(nimDisableCertificateValidation):
```
Comment (zerbina) [inline lib/wrappers/openssl.nim:799]: Is the procedures below this point becoming available on Windows intentional? If so, could you elaborate on the reason behind it?

**Review by zerbina — COMMENTED** [review]
I left a question regarding the removal of the `not defined(windows)` condition, but otherwise this looks good to me.

File: lib/wrappers/openssl.nim (hunk 1 above)
Comment (alaviss) [reply lib/wrappers/openssl.nim:799]: The reason is that there is no reason for them not to be available.

For mainline Nim, they are hidden away due to how Nim used to ship old OpenSSL 1.0 DLLs, which doesn't have these symbols, and was never enabled even after they upgraded to 1.1.1.

I'll add this to the PR description.

**Review by zerbina — APPROVED** [review]

**Pull request merged**